graphix/pattern.py

```diff
@@ -93,34 +90,98 @@ def add(self, cmd):
         assert type(cmd) == list
         assert cmd[0] in ["N", "E", "M", "X", "Z", "S", "C"]
         if cmd[0] == "N":
-            self.Nnode += 1
-            self.output_nodes.append(cmd[1])
+            if cmd[1] not in self.__output_nodes:
```

Comment:
For the time being, we accept to reprepare with an `N` command a node that is already prepared (either by a previous `N` command or by being listed in `input_nodes`): it seems preferable to fail in this case, but this breaks other tests. We should discuss whether we address this question here or for a future pull request. Ideally, we would prefer that `input_nodes` are prepared by the context (in our case, the simulator) and not by `N` commands.

Reply:
Fixed in b7985a4